### DISCUSSION TITLE
Module/Class Brainstorm

### DISCUSSION THREAD

**Discussion by blasecapital** [Blase-Labs/blase#2] [Ideas]
Let's plan the library's core modules/classes to support end-to-end model training!

Prerequisites:
The workflow begins with raw data. Users should collect and store the data in some digital way-the library should accommodate a variety of raw data sources and load them in chunks for feature/target creation.

Core:
The following are the foundational objects in the library. They should be universally applicable to deep learning pipelines.

- Extract - batch-load data from any source (assumes user did some work to make raw data accessible, such as in tabular format in .csv/.txt/.db files, pictures/text/other stored logically in a directory)
- Transform - facilitate custom functions to transform data in a way that is staged for loading
- Load - store transformed data
- Examine- describe and visualize data to inform user if cleaning is needed
- Clean - handle unwanted data, make sure that new missing data is handled appropriately for aligned datasets
- Prepare - chunk, split clean data, and convert into numpy datatype in a binary file so the dataset is ready for training and in an efficient I/O format
- Simulate - functions related to RL setup
- Train - support training in TensorFlow and PyTorch, include k-fold and other training functions, store metadata for reproducibility
- Evaluate - load models, save predictions, run metrics, explain model, enable custom functions to determine usability based on stored model outcome
- Deploy - make the model/dependencies production ready
- Monitor - post-deployment monitoring for drift detection, retraining triggers, and performance tracking
- Update - load the model to learn on new data

I am mostly familiar with supervised classification projects. The library ought to accommodate unsupervised, reinforcement learning, multi-modal, and fine-tuning projects. Please jump in to share what is missing and what can be optimized in the comments!

**Comment by blasecapital** [comment]
Dev note: design classes to be accommodative with default, configurable, and extendable settings.

**Comment by blasecapital** [comment]
# Managing Memory, Storage, and Logging in `blase`

The `blase` framework is designed to support **local-first deep learning workflows** with arbitrarily large datasets. To avoid memory bloat while maintaining reproducibility, it uses **batch-wise processing**, **hash-based logging**, and **optional storage of intermediate datasets**.

---

## Data Flow in `blase`

```
raw_data/        ← loaded by Extract
    ↳ cleaned_data/     ← optional output after Clean
        ↳ prepped_data/ ← binary training-ready data from Prepare
            ↳ predictions/, logs/, models/ ← outputs from Train/Evaluate
```

---

## Storage Recommendations by Stage

| Stage              | Store to Disk? | Reason                                                      |
|-------------------|----------------|-------------------------------------------------------------|
| `raw_data/`        | Yes         | Source of truth                                             |
| `cleaned_data/`    | Optional   | Save only if cleaning is expensive or irreversible         |
| `prepped_data/`    | Yes         | Required for training and evaluation                        |
| Intermediate batches | No        | Use memory or temp files, not permanent storage             |
| `logs/`, `hashes/` | Yes         | Required for reproducibility and caching                    |
| `models/`, `predictions/` | Yes | Versioned outputs                                           |

---

## Why Hashing and Logging Matter

Logging and hashing allow `blase` to:

- Track what raw data + transform function produced what output
- Avoid storing every file permanently
- Allow users to **rebuild** intermediate stages without resaving them
- Enable reproducibility with lightweight metadata

### Example Log Record
```json
{
  "input_hash": "abc123",
  "transform_function_hash": "xyz987",
  "output_path": "prepped/v1.npy",
  "timestamp": "2025-03-21T12:00:00Z"
}
```

---

## Strategies to Prevent Memory Bloat

| Strategy                     | Implementation                                                   |
|-----------------------------|-------------------------------------------------------------------|
| Optional Checkpointing      | Users can opt to skip saving `cleaned_data/` or `transform_output/` |
| Smart Hash-Based Logging    | Store hashes of input + transform to recreate data when needed     |
| Cache Cleanup Utility       | CLI command to purge temp/intermediate files while keeping logs     |
| Lazy Batching & Iterators   | Use `yield`/generators to avoid memory overflow                     |
| Configurable Cache Depth    | Let users define how much data to keep (`--cache-depth`, etc.)      |

---

## Suggested Best Practices

- **Store only:**
  - `raw_data/`
  - `prepped_data/`
  - `models/`, `logs/`, and versioned outputs

- **Recompute everything else** from:
  - Hash logs
  - Custom functions (stored as `.py` or via import path)

- **Use CLI tools** to:
  - Re-run entire pipelines from raw with one command
  - Simulate a run (`--dry-run`) to verify reproducibility
  - Clean unused intermediate data

---

## Related `blase` Modules

- `Extract`: Loads raw or prepared data in memory-efficient batches
- `Transform`: Applies hash-logged feature engineering or cleaning
- `Prepare`: Stores training-ready `.npy` or `.tfrecord` files
- `Logger`: Centralized hashing, tagging, and reproducibility tracker

---

This memory-conscious lifecycle keeps `blase` lightweight, efficient, and reproducible—even when working with datasets too large to fit in memory.

**Comment by blasecapital** [comment]
# Blase Tracking System

A lightweight, reproducible experiment tracking system designed for ML/ETL pipelines.

This tracking module ensures **reproducibility, introspection, and efficient reuse** by logging every pipeline step as a structured artifact, capturing hashes of data, functions, and parameters. It avoids redundant computation, supports rollback, and enables traceability from raw input to final outputs.

---

## Core Philosophy

- Each **step** in the pipeline is logged as a uniquely identifiable file.
- Every step includes:
  - Parameters
  - Function reference + source hash
  - Inputs (data hash, file hash, etc.)
  - A hash digest (`step_hash`) that identifies the operation
  - Optional result reference (e.g. output hash, model ID)
  - List of **upstream step hashes** used as dependencies
- Repeated steps are detected and **replayed**, not duplicated.
- States are **composed of all upstream steps**, enabling quick re-execution.
- Function and data artifacts may be snapshotted for rollback even if deleted from the user’s working directory.

---

## Directory Structure

Each run (under `/runs/<timestamp>/`) includes:

```
runs/
└── 2025-04-05T14-22-50/
    ├── logs/
    │   ├── step_001_read_csv.json
    │   ├── step_002_transform.json
    │   └── step_003_train.json
    ├── assets/
    │   └── data_001.csv
    ├── lookup/
    │   ├── step_index.json
    │   ├── hash_to_step.json
    │   └── tags.json
    ├── dependencies.txt
    ├── run_metadata.json
    └── active_run.json
```

---

## Step Structure

Each step log (e.g., `step_003_train.json`) contains:

```json
{
  "step_id": "step_003_train",
  "function": "train_model",
  "params": {
    "epochs": 5,
    "learning_rate": 0.001
  },
  "inputs": {
    "train_data": "abc123",             // hash of upstream output
    "func_hash": "def456"
  },
  "step_hash": "abc456...",
  "status": "completed",
  "timestamp_start": "...",
  "timestamp_end": "...",
  "replays": [],
  "result": "model_hash_001"
}
```

---

## Hash-Based Deduplication

Steps are uniquely identified using:

- Function source code hash (`hash_function()`)
- Parameters passed
- Input files / object hashes

This allows the system to detect:
- Has this function already run with the same inputs?
- Is the step redundant or a new experiment?

---

## Key Functions

### Run Setup

- `Track._validate_run_directory(track: bool) -> Path`
  - Ensures a `/runs` directory exists.
  - Returns the base path.

- `Track._validate_active_run(track_dir: Path) -> Path`
  - Initializes a new run folder if none exists.
  - Returns the path to the active run directory.

- `Track.__initialize_run_structure(run_path: Path)`
  - Creates `/logs`, `/assets`, `/lookup`, etc.
  - Saves `run_metadata.json` and `dependencies.txt`.

---

### Step Lifecycle

- `Track._start_step(run_path, function, params, inputs) -> (step_id, step_file_path)`
  - Creates or reuses a tracked step file.
  - Computes a `step_hash` using function, params, and inputs.
  - Updates `hash_to_step.json` and `step_index.json`.

- `Track._end_step(step_file_path, status="completed", outputs={})`
  - Updates the step file with final status, output hashes, and end timestamp.

---

### Hash Utilities

From `Hash` class:

- `hash_function(func) -> str`
- `hash_object(obj) -> str`
- `hash_file(file_path) -> str`
- `hash_directory(dir_path) -> str`
- `compare_hashes(h1, h2) -> bool`

Used to generate stable, reproducible references to pipeline components.

---

## Example Workflow

```python
extractor = Extract()
transformer = Transform()
loader = Load()

for batch in extractor.read_csv("data.csv"):
    transformed = transformer.apply_function(batch, my_transform)
    loader.save_to_csv(transformed, "runs/.../assets/output.csv")
```

Each step (`read_csv`, `apply_function`, `save_to_csv`) is tracked, hashed, and stored in `/logs`.

---

## Step Dependency Tracing

- Each step stores a reference to its upstream hashes.
- Steps may be replayed if prior hashes match.
- Future system design may include a dependency graph or forward-replay logic for recovering deleted data.

---

## Planned Extensions

- `step_graph.json` to link steps in DAG structure
- `restore_step(step_hash)` to recreate lost data
- CLI integration for `blase track` commands
- Git-based source versioning and linkage
- Merkle-style dependency tracing (optional)

---

## Blase Asset Storage Strategy

This table outlines how `blase` handles different types of artifacts during tracking, based on their size, importance to reproducibility, and whether they should be stored or simply referenced.

| **Asset Type**           | **Tracked (Hash)** | **Stored by Default** | **Rationale**                                                                 |
|--------------------------|--------------------|------------------------|-------------------------------------------------------------------------------|
| Function source        | Yes                | Yes                    | Lightweight and critical for logic-level reproducibility.                     |
| Step metadata          | Yes                | Yes                    | Core tracking component: parameters, inputs, outputs, timestamps, etc.       |
| Workflow dependencies  | Yes                | Yes                    | Enables backward/forward lineage and dependency resolution.                   |
| Scripts (e.g. `main.py`)| Yes               | Optional               | Optional snapshot of CLI/code-run logic (useful for postmortem review).      |
| Input files (CSV, NPY) | Yes                | No                     | Tracked via hash + path; user is expected to manage or rerun to regenerate.   |
| Intermediate DataFrames| Yes (after step)   | No                     | Reproducible by rerunning previous steps; expensive to store every version.   |
| Model files / weights  | Yes                | No (optional config)   | Reference only; allow user to toggle persistence based on importance.         |
| External modules       | No                 | No                     | Assumed to be installed via package manager; not managed by `blase`.          |

> **Note:** Users can customize storage behavior in a future `blase_config.yaml` or `track_config.json` file.

### Example Configuration (coming soon):

```yaml
store_models: false
store_input_files: false
store_function_sources: true
store_scripts: true
```

This setup allows `blase` to provide strong reproducibility guarantees while minimizing disk usage.

---

## Summary

This system provides a lightweight and flexible tracking mechanism for any machine learning or data transformation pipeline, supporting reproducibility and experimentation without vendor lock-in or bulky external tools.

```bash
blase track run
blase track list
blase track restore <step_hash>
```

# No Duplication Strategy in Blase

This document outlines how **Blase** prevents redundant duplication of data, steps, and logs in deterministic workflows — particularly during ETL and preprocessing steps — while allowing controlled duplication for inherently **non-deterministic processes** like model training.

---

## Goals

- Prevent multiple storage of **identical inputs**, **outputs**, and **functions**.
- Enable **reproducibility** without storing large files multiple times.
- Allow **re-runs** of stochastic processes (e.g., training) while still logging uniquely.
- Reduce unnecessary **I/O overhead** and disk usage.

---

## Key Concepts

| Concept              | Description                                                                 |
|----------------------|-----------------------------------------------------------------------------|
| **Step Hash**        | Each step is hashed from `{function_name, params, input_hashes}`.            |
| **Replays**          | When a step is re-executed with the same inputs, its `step_#.json` is reused and `timestamp` is added to a `replays` list. |
| **Asset Deduplication** | Files (e.g. CSVs, NumPy arrays) are hashed; if a hash exists, they are not duplicated. |
| **Function Hashing** | Functions are hashed using introspection; identical logic is not re-logged. |

---

## Deduplication Rules

### 1. **ETL Steps (Deterministic)**

These steps **should not be duplicated** unless the code or input data changes.

| Stage       | Duplicated? | Reason                                                       |
|-------------|-------------|--------------------------------------------------------------|
| `Extract`   | ❌           | Hashes file + params. If unchanged, logs a replay.          |
| `Transform` | ❌           | Hashes function + batch + args. If unchanged, reuses step.  |
| `Load`      | ❌           | Checks if file already exists with identical content hash.  |

### 2. **Model Training (Stochastic)**

These steps **can be duplicated**, since the results are inherently non-deterministic (e.g., due to random seeds, dropout, etc.).

| Stage    | Duplicated? | Reason                                                         |
|----------|-------------|----------------------------------------------------------------|
| `Train`  | ✅           | Even with same data and code, random seeds yield new results.  |
| `Evaluate` | ✅        | Evaluation tied to trained model version.                      |

---

## Example: Data Step Reuse

```json
{
  "step_id": "step_004_read_csv",
  "function": "read_csv",
  "inputs": {
    "file_hash": "abc123",
    "params": {"backend": "polars"}
  },
  "step_hash": "xyz789",
  "replays": ["2025-04-05T13:15:01", "2025-04-06T09:45:22"]
}
```

Only one `step_004_read_csv.json` created  
`replays` updated on subsequent calls with identical config

---

## How It Works

- Before saving a step, **Blase computes the step hash**
- Checks if `hash_to_step.json` already contains this hash:
  - If yes → Append to `replays`, skip asset copy
  - If no → Create new `step_#.json`, log inputs, optionally store assets
- Applies same check for **output assets** (e.g. using `hash_file` or `hash_object`)

---

## What Isn't Duplicated

- Identical scripts or notebooks
- Same `.csv`, `.npy`, `.json` batch files
- Repeated `read_csv()` or `transform()` operations with same function

---

## What *Can* Be Duplicated (If Needed)

- Training steps with random weights/augments
- A step explicitly marked with `force=True`
- Intermediate assets if the user wants a snapshot at every run

---

## Optional Enhancements

- `config.yml` field: `enforce_no_duplication: true`
- Per-step override: `save_to_csv(..., force_save=True)`
- Diff utility to compare inputs between steps

---

## Summary

Blase ensures **reproducible**, **efficient**, and **minimal** storage by default.  
Redundant actions are deduplicated intelligently, while allowing for variation where necessary.

```python
# Deterministic (deduplicated)
transformer.apply_function(batch, custom_cleaning_func)

# Stochastic (tracked as new step)
model.train(X_train, y_train, seed=random())
```

---
